include/rtio.h

```diff
@@ -0,0 +1,479 @@
+	u32_t begin_tstamp;
+
+	/** Timestamp in cycles from k_cycle_get_32() marking when a read or
+	 * write was complete
+	 */
+	u32_t end_tstamp;
+
+	/** Byte layout designator given by each driver to specify
+	 * the layout of the contained buffer
+	 */
+	u32_t layout;
+
+	/** Length of data in this buffer. */
+	u16_t len;
+
+	/** Amount of data this buffer can store. */
+	u16_t size;
+
+	/** Start of the buffer data .*/
+	u8_t *_data;
+};
+
+/**
+ * @brief Initialize an rtio_block
+ */
+__syscall void rtio_block_init(struct rtio_block *block,
+				   struct k_mem_block_id mem_block_id,
+			       u8_t *data, u32_t size);
+
+static inline void z_impl_rtio_block_init(struct rtio_block *block,
+				   struct k_mem_block_id mem_block_id,
+				   u8_t *data, u32_t size)
+{
+	block->mem_block_id = mem_block_id;
+	block->layout = 0;
+	block->begin_tstamp = 0;
+	block->end_tstamp = 0;
+	block->len = 0;
+	block->size = size;
+	block->_data = data;
+}
+
+/**
+ * @brief Begin a block read or write
+ */
+__syscall void rtio_block_begin(struct rtio_block *block);
+
+static inline void z_impl_rtio_block_begin(struct rtio_block *block)
+{
+	block->len = 0;
+	block->begin_tstamp = k_cycle_get_32();
+}
+
+/**
+ * @brief End a block read or write.
+ */
+__syscall void rtio_block_end(struct rtio_block *block);
+
+static inline void z_impl_rtio_block_end(struct rtio_block *block)
+{
+	block->end_tstamp = k_cycle_get_32();
+}
+
+
+/**
+ * @brief Unused number of bytes in the buffer
+ */
+__syscall u16_t rtio_block_available(struct rtio_block *block);
+
+static inline u16_t z_impl_rtio_block_available(struct rtio_block *block)
+{
+	return block->size - block->len;
+}
+
+/**
+ * @private
+ * @brief Implement a push/pull function pair
+ *
+ * Implements a pair of functions for pushing or pulling a fixed size value
+ * into an rtio block.
+ */
+#define Z_RTIO_BLOCK_PUSH_PULL_IMPL(_type)				\
+	static inline int rtio_block_push_##_type(			\
+		struct rtio_block *block, _type val)			\
+	{								\
+		if (rtio_block_available(block) >= sizeof(_type)) {	\
+			memcpy(&block->_data[block->len], &val,		\
+			       sizeof(_type));				\
+			block->len += sizeof(_type);			\
+			return 0;				        \
+		} else {						\
+			return -ENOMEM;					\
+		}							\
+	}								\
+									\
+	static inline int rtio_block_pull_##_type(			\
+		struct rtio_block *block, u16_t offset, _type * val)	\
+	{								\
+		if (block->len - offset >= sizeof(_type)) {		\
+			memcpy(val, &block->_data[offset],		\
+			       sizeof(_type));				\
+			return offset + sizeof(_type);			\
+		} else {						\
+			return -EINVAL;					\
+		}							\
+	}
+
+/**
+ * @brief Push/pull a value in/out of the block
+ *
+ * Implementations for common data types you might wish to
+ * store in a rtio_block
+ */
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(u8_t)
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(u16_t)
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(u32_t)
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s8_t)
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s16_t)
+Z_RTIO_BLOCK_PUSH_PULL_IMPL(s32_t)
```

Comment:
Do you have any thoughts on 64 bit support?

I'm not asking for 64 bit support right now -- I'm just suggesting it should be seriously considered, since Zephyr is being made 64-bit safe.

Reply:
I added s64/u64 types here